devito/passes/clusters/asynchrony.py

```diff
@@ -1,326 +1,316 @@
+
+    def memcpy_key(c):
+        if task_key(c):
+            # Writes would take precedence over reads
+            return None
+        else:
+            return async_trigger(c, key0(c.scope.reads))
 
+    return task_key, memcpy_key
 
-class Tasker(Asynchronous):
 
+@timed_pass(name='tasking')
+def tasking(clusters, key0, sregistry):
+    """
+    Turn a Cluster `c` into an asynchronous task if `c` writes to a Function `f`
+    such that `key0(f) = (d_i, ..., d_n)`, where `(d_i, ..., d_n)` are the
+    task Dimensions.
     """
-    Create asynchronous Clusters, or "tasks".
+    key, _ = keys(key0)
+    return Tasking(key, sregistry).process(clusters)
 
-    Parameters
-    ----------
-    key : callable, optional
-        A Cluster `c` becomes an asynchronous task only if `key(c)` returns True
 
-    Notes
-    -----
-    From an implementation viewpoint, an asynchronous Cluster is a Cluster
-    with attached suitable SyncOps, such as WaitLock, WithLock, etc.
+class Tasking(Queue):
+
+    """
+    Carry out the bulk of `tasking`.
     """
 
-    @timed_pass(name='tasker')
-    def process(self, clusters):
-        return super().process(clusters)
+    def __init__(self, key0, sregistry):
+        self.key0 = key0
+        self.sregistry = sregistry
 
     def callback(self, clusters, prefix):
         if not prefix:
             return clusters
 
-        d = prefix[-1].dim
+        dim = prefix[-1].dim
 
         locks = {}
-        waits = defaultdict(OrderedSet)
-        tasks = defaultdict(list)
+        syncs = defaultdict(lambda: defaultdict(OrderedSet))
         for c0 in clusters:
-            dims = self.key(c0)
-            if d not in dims:
-                # Not a candidate asynchronous task
+            d = self.key0(c0)
+            if d is not dim:
                 continue
 
-            # Prevent future writes to interfere with a task by waiting on a lock
-            may_require_lock = c0.scope.reads
-
-            # We can ignore scalars as they're passed by value
-            may_require_lock = {f for f in may_require_lock if f.is_AbstractFunction}
+            protected = self._schedule_waitlocks(c0, d, clusters, locks, syncs)
+            self._schedule_withlocks(c0, d, protected, locks, syncs)
 
-            # Sort for deterministic code generation
-            may_require_lock = sorted(may_require_lock, key=lambda i: i.name)
+        processed = [c.rebuild(syncs={**c.syncs, **syncs[c]}) for c in clusters]
 
-            protected = defaultdict(set)
-            for c1 in clusters:
-                offset = int(clusters.index(c1) <= clusters.index(c0))
+        return processed
 
-                for target in may_require_lock:
-                    try:
-                        writes = c1.scope.writes[target]
-                    except KeyError:
-                        # No read-write dependency, ignore
+    def _schedule_waitlocks(self, c0, d, clusters, locks, syncs):
```

Comment:
Note for reviewers: this is mostly the same as before (with same exceptions), but split over different methods for better code readability